coding/python_testing.rst

```diff
@@ -4,6 +4,204 @@ Python Unit Testing
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
+.. note::
+   The normal way to run tests is to use :command:`scons` or :command:`scons tests`.
+   :command:`scons` will then execute each of the tests and report any failures.
+   Currently tests are executed one at a time using :command:`python` and checking the command exit status.
+   Eventually :lmod:`~lsst.sconsUtils` will be modified to call :command:`py.test` directly with all Python test files (possibly with a different order each invocation).
+   It is therefore important that during this transition period developers check that tests run correctly with `pytest`_ by explicitly running :command:`py.test` when creating or updating test files.
+
+`pytest`_ can run tests from more than one file in a single invocation and this can be used to verify that there is no state contaminating later tests.
+To run `pytest`_ use the :command:`py.test` executable:
+
+.. code-block:: shell
+
+   $ py.test tests/*.py
+
+and to ensure that the order of test execution does not matter it is useful to sometimes run the tests in reverse order:
+
+.. code-block:: shell
+
+   $ py.test `ls -r tests/*.py`
+
+When writing tests it is important that tests are skipped using the proper :mod:`unittest` :ref:`skipping framework <python:unittest-skipping>` rather than returning from the test early.
+:mod:`unittest` supports skipping of individual tests and entire classes using decorators or skip exceptions.
+LSST code sometimes raises skip exceptions in :meth:`~unittest.TestCase.setUp` or :meth:`~unittest.TestCase.setUpClass` class methods and :mod:`unittest` provides much flexibility in how skipping can be implemented.
+It is also possible to indicate that a particular test is expected to fail, being reported as an error if the test unexpectedly passes.
+Expected failures can be used to write test code that triggers a reported bug before the fix to the bug has been implemented and without causing the continuous integration system to die.
+One of the primary advantages of using a modern test runner such as `pytest`_ is that it is very easy to generate machine-readable pass/fail/skip/xfail statistics to see how the system is evolving over time.
+
+Using a shared base class
+=========================
+
+For some tests it is helpful to provide a base class and then share it amongst multiple test classes that are configured with different attributes.
+If this is required be careful to not have helper functions prefixed with ``test`` and do not have the base class named with a ``Test`` prefix and ensure it does not inherit from :class:`~unittest.TestCase`, if you do `pytest`_ will attempt to find tests inside it.
```

Comment:
Comma before "be careful".  Replace the first "and" with a period. Add a comma between "prefix" and "and".  Replace the comma with a semicolon.  Add a comma between "do" and "pytest".